File: src/ios/epos2Plugin.m
#import <Cordova/CDV.h>
#import "ePOS2.h"
//#import "ePOSEasySelect.h"

NSString* ALIGN_LEFT = @"[LEFT]";
NSString* ALIGN_CENTER = @"[CENTER]";
NSString* ALIGN_RIGHT = @"[RIGHT]";
NSString* TEXT_BOLD = @"[BOLD]";
NSString* TEXT_UNDERLINE = @"[UNDERLINE]";
NSString* TEXT_NORMAL = @"[NORMAL]";
NSString* TEXT_SMALL = @"[SMALL]";
NSString* TEXT_MEDIUM = @"[MEDIUM]";
NSString* CUT_FEED = @"[CUT]";

@interface epos2Plugin: CDVPlugin <Epos2DiscoveryDelegate, Epos2PtrReceiveDelegate>
{
    Epos2Printer *printer;
    Epos2FilterOption *filterOption_;
    int printerSeries;
    int lang;
    NSString* discoverCallbackId_;
    NSString* printCallbackId_;
}
@end

@implementation epos2Plugin

- (void)pluginInitialize {
    printerSeries = EPOS2_TM_T88;
    lang = EPOS2_MODEL_ANK;
}

- (void)startDiscover:(CDVInvokedUrlCommand *)command {
    discoverCallbackId_ = command.callbackId;

    filterOption_ = nil;
    filterOption_ = [[Epos2FilterOption alloc] init];
    [filterOption_ setDeviceType:EPOS2_TYPE_PRINTER];
    filterOption_.portType = EPOS2_PORTTYPE_TCP;

    int result = [Epos2Discovery start:filterOption_ delegate:self];
    if (EPOS2_SUCCESS != result) {
        NSLog(@"error in startDiscover()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in discovering printer."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
    } else {
        // Wait for discovery to pick up something.
//        sleep(1000);
        
        
    }
}

- (void) onDiscovery:(Epos2DeviceInfo *)deviceInfo {
    NSLog(@"onDiscovery: %@", [deviceInfo getIpAddress]);
    NSDictionary *info = @{@"ipAddress" : [deviceInfo getIpAddress],
                           @"deviceType": [NSNumber numberWithInt:[deviceInfo getDeviceType]],
                           @"target": [deviceInfo getTarget],
                           @"deviceName": [deviceInfo getDeviceName],
                           @"macAddress": [deviceInfo getMacAddress],
                           @"bdAddress": [deviceInfo getBdAddress],
                           };
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:info];
    [result setKeepCallbackAsBool:YES];
    [self.commandDelegate sendPluginResult:result callbackId:discoverCallbackId_];
}

- (void)stopDiscover:(CDVInvokedUrlCommand *)command {
    NSLog(@"stopDiscover()");
    int result = EPOS2_SUCCESS;

    while (YES) {
        result = [Epos2Discovery stop];

        if (result != EPOS2_ERR_PROCESSING) {
            break;
        }
    }

    if (EPOS2_SUCCESS != result) {
        NSLog(@"error in stopDiscover()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in stop discovering printer."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
    }
}

-(void)connectPrinter:(CDVInvokedUrlCommand *)command {
    NSString* target = [command.arguments objectAtIndex:0];
    NSLog(@"connectPrinter(), ip=%@", target);

    printer = [[Epos2Printer alloc] initWithPrinterSeries:printerSeries lang:lang];

    [printer setReceiveEventDelegate:self];

    int result = EPOS2_SUCCESS;

    if (printer == nil) {
        NSLog(@"error in connectPrinter(), printer object not found");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in connect  printer. Printer object not found"];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }

    result = [printer connect:[NSString stringWithFormat:@"TCP:%@", target] timeout:EPOS2_PARAM_DEFAULT];
    if (result != EPOS2_SUCCESS) {
        NSLog(@"error in connectPrinter()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in connect  printer."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }

    result = [printer beginTransaction];
    if (result != EPOS2_SUCCESS) {
        NSLog(@"error in beginTransaction()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in begin transaction."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        [printer disconnect];
        return;
    }

    CDVPluginResult *cordovaResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:YES];
    [self.commandDelegate sendPluginResult:cordovaResult callbackId:command.callbackId];
}

- (void)disconnectPrinter:(CDVInvokedUrlCommand *)command {
    int result = EPOS2_SUCCESS;
    //    NSMutableDictionary *dict = [[NSMutableDictionary alloc] init];

    if (printer == nil) {
        return;
    }

    result = [printer endTransaction];
    if (result != EPOS2_SUCCESS) {
        NSLog(@"error in disconnectPrinter()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in end transaction."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        //        [dict setObject:[NSNumber numberWithInt:result] forKey:@"result"];
        //        [dict setObject:@"endTransaction" forKey:@"method"];
        //        [self performSelectorOnMainThread:@selector(showEposErrorFromThread:) withObject:dict waitUntilDone:NO];
    }

    result = [printer disconnect];
    if (result != EPOS2_SUCCESS) {
        NSLog(@"error in disconnectPrinter()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in disconnect printer."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        //        [dict setObject:[NSNumber numberWithInt:result] forKey:@"result"];
        //        [dict setObject:@"disconnect" forKey:@"method"];
        //        [self performSelectorOnMainThread:@selector(showEposErrorFromThread:) withObject:dict waitUntilDone:NO];
    }
    [self finalizeObject];

    CDVPluginResult *cordovaResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:YES];
    [self.commandDelegate sendPluginResult:cordovaResult callbackId:command.callbackId];
}

- (void)finalizeObject
{
    if (printer == nil) {
        return;
    }

    [printer clearCommandBuffer];

    [printer setReceiveEventDelegate:nil];

    printer = nil;
}

- (void)onPtrReceive:(Epos2Printer *)printerObj code:(int)code status:(Epos2PrinterStatusInfo *)status printJobId:(NSString *)printJobId {
    NSLog(@"onPtrReceive");
    NSLog(@"code: %d", code);
    NSLog(@"status: %@", status);
    NSLog(@"printJobId: %@", printJobId);

    // Again, same as on Android, why are we disconnecting here?????
    //[self disconnectPrinter:nil];
}

- (void)print:(CDVInvokedUrlCommand *)command {
    printCallbackId_ = command.callbackId;

    if (printer == nil) {
        NSLog(@"error in createPrintData()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in create data, printer object not dound."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }

    NSArray *printData = command.arguments;

    int result = EPOS2_SUCCESS;

    result += [self addLines:printData];

    result += [printer addCut:EPOS2_CUT_FEED];

    if (result != EPOS2_SUCCESS) {
        return;
    }

    [self sendDataToPrinter:command.callbackId];
}

- (void)printImage:(CDVInvokedUrlCommand *)command {
        
}

- (void)printReceipt:(CDVInvokedUrlCommand *)command {
    printCallbackId_ = command.callbackId;
    
    if (printer == nil) {
        NSLog(@"printReceipt aborted, not connected");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Not connected to a printer"];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }
    
    // Get function parameters.
    NSString* image64 = [command.arguments objectAtIndex:0];
    NSArray* text = [command.arguments objectAtIndex:1];
    
    int result = EPOS2_SUCCESS;
    
    // Add image if present
    if (![image64 isEqualToString:@""]) {
        UIImage *image = [self imageFromBase64:image64];
        
        long scale = 1.5;
        long h = (long)image.size.height * scale;
        long w = (long)image.size.width * scale;
        
        [printer addImage:image x:0 y:0 width:w height:h color:EPOS2_PARAM_DEFAULT mode:EPOS2_MODE_MONO halftone:EPOS2_HALFTONE_DITHER brightness:1.0 compress:EPOS2_PARAM_DEFAULT];
    }
    
    // Process each line of text, filtering out commands and executing them
    result += [self addLines:text];

    // Perform a feed cut at the end of the receipt.
    result += [printer addCut:EPOS2_CUT_FEED];

    if (result != EPOS2_SUCCESS) {
        NSLog(@"One of the add commands failed");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Failed to build receipt"];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }

    // Send the commands to the printer for execution.
    [self sendDataToPrinter:command.callbackId];
    
}

- (int)addLines:(NSArray*) lines
{
    int result = EPOS2_SUCCESS;
    
    for (NSString* line in lines) {
        NSLog(@"%@", line);
        if ([line isEqualToString:@"\n"] || [line isEqualToString:@""]) {
            result += [printer addFeedLine:1];
        } else if ([line isEqualToString:ALIGN_LEFT]) {
            result += [printer addTextAlign:EPOS2_ALIGN_LEFT];
        } else if ([line isEqualToString:ALIGN_CENTER]) {
            result += [printer addTextAlign:EPOS2_ALIGN_CENTER];
        } else if ([line isEqualToString:ALIGN_RIGHT]) {
            result += [printer addTextAlign:EPOS2_ALIGN_RIGHT];
        } else if ([line isEqualToString:TEXT_MEDIUM]) {
            result += [printer addTextSize:2 height:2];
        } else if ([line isEqualToString:TEXT_SMALL]) {
            result += [printer addTextSize:1 height:1];
        } else if ([line isEqualToString:TEXT_BOLD]) {
            result += [printer addTextStyle:EPOS2_FALSE ul:EPOS2_FALSE em:EPOS2_TRUE color:EPOS2_PARAM_DEFAULT];
        } else if ([line isEqualToString:TEXT_UNDERLINE]) {
            result += [printer addTextStyle:EPOS2_FALSE ul:EPOS2_TRUE em:EPOS2_FALSE color:EPOS2_PARAM_DEFAULT];
        } else if ([line isEqualToString:TEXT_NORMAL]) {
            result += [printer addTextStyle:EPOS2_FALSE ul:EPOS2_FALSE em:EPOS2_FALSE color:EPOS2_PARAM_DEFAULT];
        } else if ([line isEqualToString:CUT_FEED]) {
            result += [printer addCut:EPOS2_CUT_FEED];
        } else {
            result += [printer addText:line] + [printer addFeedLine:1];
        }

        if (result != EPOS2_SUCCESS) {
            break;
        }
    }
    
    /*
     NSString* ALIGN_LEFT = @"[LEFT]";
     NSString* ALIGN_CENTER = @"[CENTER]";
     NSString* ALIGN_RIGHT = @"[RIGHT]";
     NSString* TEXT_BOLD = @"[BOLD]";
     NSString* TEXT_UNDERLINE = @"[UNDERLINE]";
     NSString* TEXT_ITALIC = @"[UNDERLINE]";
     NSString* TEXT_NORMAL = @"[NORMAL]";
     NSString* TEXT_SMALL = @"[SMALL]";
     NSString* TEXT_MEDIUM = @"[MEDIUM]";
     NSString* CUT_FEED = @"[CUT]";
     **/
    
    return result;
}

- (void)sendDataToPrinter:(NSString*) callbackId
{
    int result = EPOS2_SUCCESS;

    if (printer == nil) {
        return;
    }

    result = [printer sendData:EPOS2_PARAM_DEFAULT];
    if (result != EPOS2_SUCCESS) {
        [printer disconnect];

        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in print data."];
        [self.commandDelegate sendPluginResult:result callbackId:callbackId];
        return;
    } else {
        // Empty the buffer, removing the printed receipt.
        [printer clearCommandBuffer];
    }

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:YES];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:callbackId];
}

- (UIImage*)imageFromBase64:(NSString*) base64Text
{
    NSData *dataEncoded = [[NSData alloc] initWithBase64EncodedString:base64Text options:0];
    return [UIImage imageWithData:dataEncoded];
}

@end

File: src/android/epos2Plugin.java
package com.cordova.plugin;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.util.Base64;
import android.util.Log;

import com.epson.epos2.Epos2Exception;
import com.epson.epos2.discovery.Discovery;
import com.epson.epos2.discovery.DiscoveryListener;
import com.epson.epos2.discovery.FilterOption;
import com.epson.epos2.printer.Printer;
import com.epson.epos2.printer.PrinterStatusInfo;
import com.epson.epos2.printer.ReceiveListener;
import com.epson.eposprint.Print;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.PluginResult;
import org.apache.cordova.PluginResult.Status;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class epos2Plugin extends CordovaPlugin {
    private static final String TAG = "EPOS2Plugin";
    private CallbackContext callbackContext = null;
    private Printer printer = null;

    private final String ALIGN_LEFT = "[LEFT]";
    private final String ALIGN_CENTER = "[CENTER]";
    private final String ALIGN_RIGHT = "[RIGHT]";

    private final String TEXT_BOLD = "[BOLD]";
    private final String TEXT_UNDERLINE = "[UNDERLINE]";
    private final String TEXT_NORMAL = "[NORMAL]";

    private final String TEXT_SMALL = "[SMALL]";
    private final String TEXT_MEDIUM = "[MEDIUM]";

    private final String CUT_FEED = "[CUT]";

    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);
    }

    public boolean execute(final String action, final JSONArray args, final CallbackContext callbackContext) {
        this.callbackContext = callbackContext;

        try {
            if (action.equals("startDiscover")) {
                startDiscovery(callbackContext);
            } else if (action.equals("stopDiscover")) {
                stopDiscovery(callbackContext);
            } else if (action.equals("connectPrinter")) {
                String target = args.getString(0);
                connectPrinter(target, callbackContext);
            } else if (action.equals("disconnectPrinter")) {
                disconnectPrinter();
            } else if (action.equals("print")) {
                String[] lines = convertToArray(args, 0);
                printLines(lines, callbackContext);
            } else if (action.equals("printImage")) {
                String imageData = args.getString(0);
                printImage(imageData, callbackContext);
            } else if (action.equals("printReceipt")) {
                String image64 = args.getString(0);
                String[] lines = convertToArray(args.getJSONArray(1), 0);
                printReceipt(image64, lines, callbackContext);
            } else if (action.equals("checkStatus")) {
                checkStatus(callbackContext);
            }
        } catch (JSONException jse) {
            callbackContext.error("Command failed due to bad parameters.");
        }

        return true;
    }

    private void startDiscovery(final CallbackContext callbackContext) {
        FilterOption mFilterOption = new FilterOption();
        mFilterOption.setDeviceType(Discovery.TYPE_PRINTER);
        mFilterOption.setEpsonFilter(Discovery.FILTER_NAME);
        try {
            Discovery.start(webView.getContext(), mFilterOption, discoveryListener);
        } catch (Epos2Exception e) {
            Log.e(TAG, "Error discovering printer: " + e.getErrorStatus(), e);
            callbackContext.error("Error discovering printer: " + e.getErrorStatus());
        }
    }

    private void stopDiscovery(final CallbackContext callbackContext) {
        while (true) {
            try {
                Discovery.stop();
                PluginResult result = new PluginResult(Status.OK, true);
                callbackContext.sendPluginResult(result);
                break;
            } catch (Epos2Exception e) {
                if (e.getErrorStatus() != Epos2Exception.ERR_PROCESSING) {
                    PluginResult result = new PluginResult(Status.ERROR, false);
                    callbackContext.sendPluginResult(result);
                    return;
                }
            }
        }
    }

    private void connectPrinter(final String targetIP, final CallbackContext callbackContext) {
        Log.d(TAG, "Connecting Printer");

        try {
            printer = new Printer(Printer.TM_T88, Printer.MODEL_ANK, webView.getContext());
            printer.setReceiveEventListener(receiveListener);
        } catch (Epos2Exception e) {
            callbackContext.error("Error creating printer: " + e.getErrorStatus());
            Log.e(TAG, "Error creating printer: " + e.getErrorStatus(), e);
            return;
        }

        try {
            printer.connect("TCP:" + targetIP, Printer.PARAM_DEFAULT);
        } catch (Epos2Exception e) {
            callbackContext.error("Error connecting printer: " + e.getErrorStatus());
            Log.e(TAG, "Error connecting printer: " + e.getErrorStatus(), e);
            return;
        }

        try {
            printer.beginTransaction();
        } catch (Epos2Exception e) {
            callbackContext.error("Error beginning transaction");
            Log.e(TAG, "Error beginning transaction", e);
            return;
        }

        PluginResult result = new PluginResult(Status.OK, "Done connecting");
        callbackContext.sendPluginResult(result);
    }

    private void disconnectPrinter() {
        if (printer == null) {
            callbackContext.success("already disconnected");
            return;
        }

        try {
            printer.endTransaction();
        } catch (Epos2Exception e) {
            callbackContext.error("Error ending transaction: " + e.getErrorStatus());
            Log.e(TAG, "Error ending transaction: " + e.getErrorStatus(), e);
            return;
        }

        try {
            printer.disconnect();
        } catch (Epos2Exception e) {
            callbackContext.error("Error disconnecting printer: " + e.getErrorStatus());
            Log.e(TAG, "Error disconnecting printer: " + e.getErrorStatus(), e);
            return;
        }

        printer.clearCommandBuffer();
        printer.setReceiveEventListener(null);
        printer = null;

        callbackContext.success("disconnected");
    }

    private void printLines(final String[] lines, final CallbackContext mCallbackContext) {
        try {
            if (notReadyToPrint()) {
                mCallbackContext.error("Not ready to print");
                return;
            }

            for (String line : lines) {
                if ("\n".equals(line) || "".equals(line.trim())) {
                    printer.addFeedLine(1);
                } else if (line.equals(ALIGN_LEFT)) {
                    printer.addTextAlign(Printer.ALIGN_LEFT);
                } else if (line.equals(ALIGN_CENTER)) {
                    printer.addTextAlign(Printer.ALIGN_CENTER);
                } else if (line.equals(ALIGN_RIGHT)) {
                    printer.addTextAlign(Printer.ALIGN_RIGHT);
                } else if (line.equals(TEXT_BOLD)) {
                    printer.addTextStyle(Printer.FALSE, Printer.FALSE, Printer.TRUE, Printer.PARAM_DEFAULT);
                } else if (line.equals(TEXT_UNDERLINE)) {
                    printer.addTextStyle(Printer.FALSE, Printer.TRUE, Printer.FALSE, Printer.PARAM_DEFAULT);
                } else if (line.equals(TEXT_NORMAL)) {
                    printer.addTextStyle(Printer.FALSE, Printer.FALSE, Printer.FALSE, Printer.PARAM_DEFAULT);
                } else if (line.equals(TEXT_SMALL)) {
                    printer.addTextSize(1, 1);
                } else if (line.equals(TEXT_MEDIUM)) {
                    printer.addTextSize(2, 2);
                } else if (line.equals(CUT_FEED)) {
                    printer.addCut(Printer.CUT_NO_FEED);
                } else {
                    printer.addText(line);
                    printer.addFeedLine(1);
                }

            }

            printer.addFeedLine(2);

            printer.addCut(Printer.CUT_FEED);

            printer.sendData(Printer.PARAM_DEFAULT);

            Thread.sleep(300);
            printer.clearCommandBuffer();
        } catch (Epos2Exception e) {
            mCallbackContext.error("Printing failed due to error: " + e.getErrorStatus());
        } catch (InterruptedException e) {
            mCallbackContext.error("Printing failed because the thread was interrupted.");
        }
    }

    private void printImage(final String imageBase64, final CallbackContext mCallbackContext) {
        try {
            if (notReadyToPrint()) {
                mCallbackContext.error("Not ready to print");
//
//                // Attempt to disconnect from printer
//                if (!rawDisconnect()) {
//                    mCallbackContext.error("Error disconnecting");
//                    Log.e(TAG, "Error disconnecting");
//                }
                return;
            }

            addImage(imageBase64, 512, 100);

            printer.sendData(Printer.PARAM_DEFAULT);

            // Wait for a little bit, then clear buffer to avoid double prints
            Thread.sleep(200);
            printer.clearCommandBuffer();

            mCallbackContext.success("Printing completed");
        } catch (Epos2Exception e) {
            mCallbackContext.error("Printing failed due to error: " + e.getErrorStatus());
        } catch (InterruptedException e) {
            mCallbackContext.error("Printing failed because the thread was interrupted.");
        }
    }

    private void printReceipt(final String imageBase64, final String[] lines, final CallbackContext mCallbackContext) {
        try {
            if (notReadyToPrint()) {
                mCallbackContext.error("Not ready to print");
                return;
            }

            // Add image if present
            addImage(imageBase64, 512, 100);

            // Print text along with rest of buffer
            printLines(lines, mCallbackContext);

            mCallbackContext.success("Printing complete");

        } catch (Epos2Exception e) {
            mCallbackContext.error("Printing failed due to error: " + e.getErrorStatus());
        }
    }

    private void addImage(final String imageBase64, int maxWidth, int maxHeight) throws Epos2Exception {
        if (imageBase64 != null && imageBase64.trim() != "") {
            Bitmap image = scale(convertBase64ToBitmap(imageBase64), maxWidth, maxHeight);
            printer.addImage(image, 0, 0, image.getWidth(), image.getHeight(), Printer.PARAM_DEFAULT,
                    Printer.MODE_MONO, Printer.HALFTONE_DITHER, 1.0,
                    Printer.PARAM_DEFAULT);
        }
    }

    private void checkStatus(final CallbackContext mCallbackContext) {
        try {
            if (printer == null) {
                mCallbackContext.error("Not connected to printer");
                return;
            }

            PrinterStatusInfo stat = printer.getStatus();

            if (stat == null) {
                mCallbackContext.error("Failed to get status");
                return;
            }

            JSONObject responseObj = new JSONObject();

            responseObj.put("connected", stat.getConnection() == Printer.TRUE ? "true" : "false");
            responseObj.put("online", stat.getOnline() == Printer.TRUE ? "true" : "false");
            responseObj.put("paper", getPaperState(stat.getPaper()));
            responseObj.put("cover", stat.getCoverOpen() == Printer.TRUE ? "Open" : "Closed");

            mCallbackContext.success(responseObj);
        } catch (JSONException exc) {
            mCallbackContext.error("Failed to build status response.");
            Log.e(TAG, "Failed to build status response", exc);
        }
    }

    /**
     * Triggers when a printer is found.
     */
    private DiscoveryListener discoveryListener = deviceInfo -> {
        JSONObject item = new JSONObject();

        try {
            item.put("deviceName", deviceInfo.getDeviceName());
            item.put("target", deviceInfo.getTarget());
            item.put("ipAddress", deviceInfo.getIpAddress());
            item.put("macAddress", deviceInfo.getMacAddress());
            item.put("deviceType", deviceInfo.getDeviceType());
            item.put("bdAddress", deviceInfo.getBdAddress());
        } catch (JSONException e) {
            callbackContext.error("Error building device info");
        }

        callbackContext.success(item);
    };

    /**
     * Triggers when the printer acknowledges having received commands.
     */
    private ReceiveListener receiveListener = (printer, code, status, printJobId) -> {
        Log.d(TAG, "print success. status: " + status.getErrorStatus());

        // Disconnect after printer receives data????? why???
//            new Thread(new Runnable() {
//                @Override
//                public void run() {
//                    disconnectPrinter();
//                }
//            }).start();
    };

    private boolean notReadyToPrint() {
        if (printer == null) {
            return true;
        }

        PrinterStatusInfo status = printer.getStatus();
        if (status == null) {
            return true;
        }

        if (status.getConnection() == Printer.FALSE) {
            return true; // Not connected
        } else if (status.getOnline() == Printer.FALSE) {
            return true; // Not online (???)
        }

        return false;
    }

    private Bitmap convertBase64ToBitmap(String b64) {
        byte[] imageAsBytes = Base64.decode(b64.getBytes(), Base64.DEFAULT);
        return BitmapFactory.decodeByteArray(imageAsBytes, 0, imageAsBytes.length);
    }

    private Bitmap scale(Bitmap bitmap, int maxWidth, int maxHeight) {
        // Determine the constrained dimension, which determines both dimensions.
        int width;
        int height;
        float widthRatio = (float) bitmap.getWidth() / maxWidth;
        float heightRatio = (float) bitmap.getHeight() / maxHeight;
        // Width constrained.
        if (widthRatio >= heightRatio) {
            width = maxWidth;
            height = (int) (((float) width / bitmap.getWidth()) * bitmap.getHeight());
        }
        // Height constrained.
        else {
            height = maxHeight;
            width = (int) (((float) height / bitmap.getHeight()) * bitmap.getWidth());
        }
        Bitmap scaledBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);

        float ratioX = (float) width / bitmap.getWidth();
        float ratioY = (float) height / bitmap.getHeight();
        float middleX = width / 2.0f;
        float middleY = height / 2.0f;
        Matrix scaleMatrix = new Matrix();
        scaleMatrix.setScale(ratioX, ratioY, middleX, middleY);

        Canvas canvas = new Canvas(scaledBitmap);
        canvas.setMatrix(scaleMatrix);
        canvas.drawBitmap(bitmap, middleX - bitmap.getWidth() / 2, middleY - bitmap.getHeight() / 2, new Paint(Paint.FILTER_BITMAP_FLAG));
        return scaledBitmap;
    }

    private String getPaperState(int code) {
        switch (code) {
            case Printer.PAPER_OK:
                return "OK";
            case Printer.PAPER_NEAR_END:
                return "Low";
            case Printer.PAPER_EMPTY:
                return "Empty";
        }
        return "Unknown";
    }

    private boolean rawDisconnect() {
        try {
            if (printer == null) {
                return true;
            }
            printer.disconnect();
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    private String[] convertToArray(final JSONArray args, int startIndex) throws JSONException {
        if (startIndex > args.length()) {
            startIndex = 0;
        }
        String[] arr = new String[args.length()];
        for (int index = startIndex; index < args.length(); index++) {
            arr[index] = args.getString(index).trim();
        }
        return arr;
    }

}
